Parse URL s= parameter as point style

kw217 opened the pull request:
As outlined in #238:
* Introduce StyledPoint class, and move style handling from Bookmark up to it.
* Make ApiMarkPoint a StyledPoint.
* Parse s= optional parameter in URL as the style.

Before and after shot below, for the same URL. The uppermost point is not styled (so defaults to red on the right). The original code (on left) ignores the style parameter and always shows a pink circle. After this PR (on right) the style parameter controls the appearance of the place marker.

![omim-add-style](https://cloud.githubusercontent.com/assets/587036/10652691/8e036662-7850-11e5-88a0-a5d0eb8d462d.png)

Existing tests all pass. Which tests should I be modifying to regression-test this fix?

mapsmetest commented:
Jenkins asks: can anyone have a look at this request?
If you want Jenkins to compile and test this, write "test this please" or JT_TP (without the underscore in the middle).

kw217 commented:
test this please

ExMix commented on map/styled_point.cpp:
```diff
@@ -0,0 +1,54 @@
+
+graphics::DisplayList * StyledPoint::GetDisplayList(UserMarkDLCache * cache) const
+{
+  return cache->FindUserMark(UserMarkDLCache::Key(GetStyle(), graphics::EPosAbove, GetContainer()->GetDepth()));
```
Original ApiPoint mark should be drawed with graphics::ECenter. With EPosAbove pixelPoint{Sprite.Width / 2, Sprite.Height} will be placed on Org point and original Api style will be a little upper that it's need.

kw217 replied:
Original API style is retired by this PR. I think all points are drawn in correct location, as shown above. Unstyled points are drawn as placemark-red now. So I think this code is correct.

kw217 replied:
Sorry, my comment was a bit brief! What I mean is, my change means that the original ApiPoint mark is never drawn. One of the placemarks is always used instead. And the placemark drawing code is literally moved from the bookmark (KML) code, without significant change, so I think it ought to be right.

@mapsmetest How can I see the results of Jenkins tests?

deathbaba replied:
If you change default API behavior to draw bookmarks instead of violet API points, it will break UI in already existing apps.
It should use “default” old style if this parameter (“s”) is omitted from the API call.

kw217 replied:
OK, I'll sort that.

deathbaba commented on map/api_mark_container.cpp:
```diff
@@ -0,0 +1,29 @@
+{
+  return new ApiMarkPoint(ptOrg, this);
+}
+
```
Extra \n

deathbaba commented on map/api_mark_container.cpp:
```diff
@@ -0,0 +1,29 @@
+
+#include "framework.hpp"
+
+////////////////////////////////////////////////////////////////////////
```
This comment line is not necessary.

deathbaba commented on map/styled_point.cpp:
```diff
@@ -0,0 +1,54 @@
+
+m2::PointD const & StyledPoint::GetPixelOffset() const
+{
+  static m2::PointD s_offset(0.0, 3.0);
```
m2::PointD const

deathbaba commented on map/styled_point.cpp:
```diff
@@ -0,0 +1,54 @@
+
+  string GetDefaultStyle()
+  {
+    return s_arrSupportedColors[0];
```
Default style should be existing API point.

deathbaba commented on map/styled_point.cpp:
```diff
@@ -0,0 +1,54 @@
+  string GetSupportedStyle(string const & s, string const & context)
+  {
+    if (s.empty())
+      return s_arrSupportedColors[0];
```
return GetDefaultStyle()

deathbaba commented on map/styled_point.cpp:
```diff
@@ -0,0 +1,54 @@
+
+#include "std/fstream.hpp"
+#include "std/algorithm.hpp"
+#include "std/auto_ptr.hpp"
```
Do we really need all these includes?

kw217 replied:
D'oh, sorry. I'll trim.

deathbaba commented on map/api_mark_point.hpp:
```diff
@@ -0,0 +1,50 @@
+#include "map/user_mark.hpp"
+#include "map/styled_point.hpp"
+
+class ApiMarkPoint : public StyledPoint
```
What is the reason to have 2 separate classes for StyledPoint and ApiMarkPoint?
I see it as "extending" existing ApiMarkPoint implementation (we don't need old one, only the basic icon if style was not specified).

kw217 replied:
Because a Bookmark is a StyledPoint but not an ApiMarkPoint. StyledPoint is the common ABC of both.

deathbaba replied:
Ok.

ygorshenin replied:
Do we really need StylePoint base class for Bookmark and ApiMarkPoint? Where it's used?

ygorshenin replied:
Ah, I see: StyledPoint is a subclass of ICustomDrawable. OK.

deathbaba commented:
@kw217 How's going? Do you need any help with fixes outlined in comments?

kw217 commented:
Hopefully I'll push something tonight. Been a bit busy sorry!

kw217 commented:
OK, I think I've applied all the markups. Now a style of "" indicates to use the original point (violet circle), which is the default when used for URLs; but I've preserved the default of placemark-red for KML bookmarks.

Rendering mapswithme://map?ll=55,-3&n=alpha&id=a&s=placemark-green&ll=55,-4&n=beta&id=b&s=placemark-yellow&ll=56,-4&n=gamma&id=c&s=unknown-aardvark&ll=57,-3&n=omega&id=z now shows up as follows: for URLs, both unrecognized (Glasgow) and omitted (Aberdeen) styles show up as the original violet circle.

![take2](https://cloud.githubusercontent.com/assets/587036/10953592/b1794de6-8340-11e5-9767-5add30b5cbc4.png)

kw217 commented:
JT TP

deathbaba commented:
JTTP

deathbaba commented on map/user_mark_container.cpp:
```diff
@@ -177,6 +177,13 @@ namespace
 
+UserMarkDLCache::Key UserMarkContainer::GetDefaultKey() const
+{
+  return UserMarkDLCache::Key(GetTypeName(),
```
You can make it an one-liner.

deathbaba commented on map/styled_point.cpp:
```diff
@@ -0,0 +1,56 @@
+
+graphics::DisplayList * StyledPoint::GetDisplayList(UserMarkDLCache * cache) const
+{
+  UserMarkDLCache::Key key = GetStyle().empty() ? GetContainer()->GetDefaultKey()
```
UserMarkDLCache::Key const key

deathbaba commented on map/styled_point.cpp:
```diff
@@ -0,0 +1,56 @@
+  UserMarkDLCache::Key key = GetStyle().empty() ? GetContainer()->GetDefaultKey()
+                                                : UserMarkDLCache::Key(GetStyle(),
+                                                                       graphics::EPosAbove,
+                                                                       GetContainer()->GetDepth());
```
UserMarkContainer const * container = GetContainer();
and reuse container variable in two places here.

igrechuhin commented on map/bookmark.hpp:
```diff
@@ -80,13 +81,13 @@ class Bookmark : public ICustomDrawable
   }
 
-  void SetData(BookmarkData const & data) { m_data = data; }
+  void SetData(BookmarkData const & data) { m_data = data; SetStyle(m_data.GetType()); }
```
It should not be one line.

igrechuhin commented on map/bookmark.hpp:
```diff
@@ -96,7 +97,7 @@ class Bookmark : public ICustomDrawable
   /// @return Now its a bookmark color - name of icon file
   string const & GetType() const { return m_data.GetType(); }
-  void SetType(string const & type) { m_data.SetType(type); }
+  void SetType(string const & type) { m_data.SetType(type); SetStyle(type); }
```
It should not be one line.

gardster commented:
clang-format whole PR, please.

gardster commented on map/framework.hpp:
```diff
@@ -8,6 +8,8 @@
 
 #include "map/bookmark.hpp"
 #include "map/bookmark_manager.hpp"
+#include "map/api_mark_point.hpp"
```
sort it, please

kw217 commented:
Thanks - I've fixed all those now, I think, except for @gardster 's request to clang-format the whole PR.

When I ran clang-format (using the .clang-format in the root of omim) on just the files I've changed in this PR, it reformatted many lines, which I hadn't changed. The diffstat is below. Are you sure you want me to do that as part of this PR?

```
 map/api_mark_container.cpp  |  11 +-
 map/api_mark_container.hpp  |   2 +-
 map/api_mark_point.hpp      |  16 +-
 map/bookmark.cpp            | 681 +++++++++++++++++++++-----------------------
 map/bookmark.hpp            |  45 +--
 map/framework.hpp           | 176 ++++++++----
 map/map.pro                 | 106 ++-----
 map/mwm_url.cpp             |  11 +-
 map/mwm_url.hpp             |   2 -
 map/styled_point.cpp        |  46 ++-
 map/styled_point.hpp        |  17 +-
 map/user_mark.hpp           |  26 +-
 map/user_mark_container.cpp | 200 +++++--------
 map/user_mark_container.hpp |  19 +-
 14 files changed, 624 insertions(+), 734 deletions(-)
```

deathbaba commented:
You can do it as a separate commit.

> On 5.11.2015, at 13.52, Keith Wansbrough <notifications@github.com> wrote:
> 
> Thanks - I've fixed all those now, I think, except for @gardster 's request to clang-format the whole PR.
> 
> When I ran clang-format (using the .clang-format in the root of omim) on just the files I've changed in this PR, it reformatted many lines, which I hadn't changed. The diffstat is below. Are you sure you want me to do that as part of this PR?
> 
>  map/api_mark_container.cpp  |  11 +-
>  map/api_mark_container.hpp  |   2 +-
>  map/api_mark_point.hpp      |  16 +-
>  map/bookmark.cpp            | 681 +++++++++++++++++++++-----------------------
>  map/bookmark.hpp            |  45 +--
>  map/framework.hpp           | 176 ++++++++----
>  map/map.pro                 | 106 ++-----
>  map/mwm_url.cpp             |  11 +-
>  map/mwm_url.hpp             |   2 -
>  map/styled_point.cpp        |  46 ++-
>  map/styled_point.hpp        |  17 +-
>  map/user_mark.hpp           |  26 +-
>  map/user_mark_container.cpp | 200 +++++--------
>  map/user_mark_container.hpp |  19 +-
>  14 files changed, 624 insertions(+), 734 deletions(-)
> —
> Reply to this email directly or view it on GitHub.
>

kw217 commented:
Shall I rebase or merge from master?

deathbaba commented:
Rebase, we are trying to avoid merges.

> On 5.11.2015, at 14.16, Keith Wansbrough <notifications@github.com> wrote:
> 
> Shall I rebase or merge from master?
> 
> —
> Reply to this email directly or view it on GitHub.
>

kw217 commented:
JTTP

gardster commented:
LGTM
By the way: you can format only lines that you have changed. Just use `git clang-format master`.

kw217 commented:
Sorry, I've used the wrong author email address in the commits. I'll fix, but I won't change any of the content.

kw217 commented:
Fixed. Sorry about that. No changes to content, only author/committer.

kw217 commented:
JTTP

mpimenov commented on map/api_mark_container.cpp:
```diff
@@ -0,0 +1,18 @@
+#include "api_mark_container.hpp"
```
map/api_mark_container.hpp

kw217 replied:
OK

mpimenov commented on map/api_mark_container.cpp:
```diff
@@ -0,0 +1,18 @@
+#include "api_mark_container.hpp"
+#include "api_mark_point.hpp"
+
+#include "framework.hpp"
```
map/framework.hpp

kw217 replied:
OK.

mpimenov commented on map/api_mark_container.hpp:
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#include "user_mark_container.hpp"
```
Please add "map/" prefix and sort the includes.

kw217 replied:
OK.

mpimenov commented on map/api_mark_point.hpp:
```diff
@@ -0,0 +1,44 @@
+
+  unique_ptr<UserMarkCopy> Copy() const override
+  {
+    return unique_ptr<UserMarkCopy>(
```
make_unique

deathbaba replied:
No, we don't use C++14 yet.

> On 6.11.2015, at 02.58, mpimenov <notifications@github.com> wrote:
> 
> In map/api_mark_point.hpp:
> 
> > +               m2::PointD const & ptOrg, UserMarkContainer * container)
> > +    : StyledPoint(style, ptOrg, container), m_name(name), m_id(id)
> > +  {
> > +  }
> > +
> > +  UserMark::Type GetMarkType() const override { return UserMark::Type::API; }
> > +
> > +  string const & GetName() const { return m_name; }
> > +  void SetName(string const & name) { m_name = name; }
> > +
> > +  string const & GetID() const { return m_id; }
> > +  void SetID(string const & id) { m_id = id; }
> > +
> > +  unique_ptr<UserMarkCopy> Copy() const override
> > +  {
> > +    return unique_ptr<UserMarkCopy>(
> make_unique
> 
> —
> Reply to this email directly or view it on GitHub.
>

ygorshenin replied:
We have our own make_unique in our base library.

deathbaba replied:
Good to know, thanks.

kw217 replied:
Thanks, nice! Fixed to use that.

mpimenov commented on map/bookmark.cpp:
```diff
@@ -182,418 +156,396 @@ size_t BookmarkCategory::FindBookmark(Bookmark const * bookmark) const
-  static char const * s_arrSupportedColors[] =
+class KMLParser
+{
+  // Fixes icons which are not supported by MapsWithMe
```
We usually put a dot in the end of a comment.

kw217 replied:
OK.

mpimenov commented on map/bookmark.cpp:
```diff
@@ -182,418 +156,396 @@ size_t BookmarkCategory::FindBookmark(Bookmark const * bookmark) const
-    "placemark-pink", "placemark-brown", "placemark-green", "placemark-orange"
-  };
+    // Remove leading '#' symbol
+    string const result = s.substr(1);
```
```ASSERT(!s.empty(), ());```

kw217 replied:
OK. I was just moving existing code, but I'm happy to improve.

mpimenov commented on map/styled_point.hpp:
```diff
@@ -0,0 +1,46 @@
+    : ICustomDrawable(ptOrg, container), m_style(style)
+  {
+  }
+
```
Please add a comment of the form
```// ICustomDrawable overrides:```

kw217 replied:
OK. I'm not familiar with this convention.

ygorshenin commented on map/api_mark_container.hpp:
```diff
@@ -0,0 +1,18 @@
+  virtual string GetActiveTypeName() const;
+
+protected:
+  virtual string GetTypeName() const;
```
Please, remove virtual and add override instead. Moreover, for groups of overridden methods (in public and protected section) add following comment:
```c++
// UserMarkContainer overrides:
Type GetType() const override { ... }
string GetActiveTypeName() const override { ... }
...
```

ygorshenin commented on map/api_mark_point.hpp:
```diff
@@ -0,0 +1,44 @@
+        new UserMarkCopy(new ApiMarkPoint(m_name, m_id, GetStyle(), m_ptOrg, m_container)));
+  }
+
+  virtual void FillLogEvent(TEventContainer & details) const override
```
Could you please remove virtual here? It's redundant in presence of override here.

ygorshenin commented on map/bookmark.cpp:
```diff
@@ -182,418 +156,396 @@ size_t BookmarkCategory::FindBookmark(Bookmark const * bookmark) const
 
 namespace
 {
+string const PLACEMARK = "Placemark";
```
As you touched this code, would you be so kind to rename this constants in accordance with our style guide, like PLACEMARK -> kPlacemark, STYLE -> kStyle, STYLE_MAP -> kStyleMap, etc. ?

ygorshenin commented on map/bookmark.cpp:
```diff
@@ -182,418 +156,396 @@ size_t BookmarkCategory::FindBookmark(Bookmark const * bookmark) const
-  };
+enum GeometryType
+{
+  UNKNOWN,
```
s/UNKNOWN/GEOMETRY_TYPE_UNKNOWN/g
s/POINT/GEOMETRY_TYPE_POINT/g
...

ygorshenin commented on map/bookmark.cpp:
```diff
@@ -182,418 +156,396 @@ size_t BookmarkCategory::FindBookmark(Bookmark const * bookmark) const
 
-    bool Push(string const & name)
+    // Remove leading '#' symbol
+    map<string, graphics::Color>::const_iterator it = m_styleUrl2Color.find(styleUrl.substr(1));
```
Please, add a dot at the end of the comment, and replace full iterator typename by auto.

ygorshenin commented on map/bookmark.hpp:
```diff
@@ -196,10 +191,10 @@ class BookmarkCategory : public UserMarkContainer
 private:
   bool m_blockAnimation;
-  typedef pair<UserMark *, shared_ptr<anim::Task> > anim_node_t;
+  typedef pair<UserMark *, shared_ptr<anim::Task>> anim_node_t;
```
Please, replace typedef by using and anim_node_t by TAnimNode.

ygorshenin commented on map/map.pro:
```diff
@@ -35,6 +36,8 @@ HEADERS += \
     user_mark_container.hpp \
     user_mark.hpp \
     user_mark_dl_cache.hpp \
+    api_mark_container.hpp \
```
Feel free to sort all *.hpp and *.cpp files lexicographically in SOURCES and HEADERS blocks.

ygorshenin commented on map/styled_point.cpp:
```diff
@@ -0,0 +1,51 @@
+  return GetStyle().empty() ? s_centre : s_offset;
+}
+
+static char const * s_arrSupportedColors[] = {
```
Please, don't use this archaic static global variables, use anonymous namespaces instead. Also
in accordance with our style guides this variable should be named kArrSupportedColors.

ygorshenin replied:
Or even kSupportedColors.

kw217 replied:
Yes, no need for pseudoHungarian...

ygorshenin commented on map/styled_point.cpp:
```diff
@@ -0,0 +1,51 @@
+  if (s.empty())
+    return fallback;
+
+  for (size_t i = 0; i < ARRAY_SIZE(s_arrSupportedColors); ++i)
```
Please, wrap loop body in curly braces.

ygorshenin commented on map/styled_point.hpp:
```diff
@@ -0,0 +1,46 @@
+  {
+  }
+
+  virtual graphics::DisplayList * GetDisplayList(UserMarkDLCache * cache) const override;
```
These virtual keywords are redudant, please, remove them.

ygorshenin commented on map/user_mark_container.cpp:
```diff
@@ -173,8 +155,13 @@ void UserMarkContainer::Clear(size_t skipCount/* = 0*/)
 {
-  static unique_ptr<PoiMarkPoint> s_selectionUserMark;
-  static unique_ptr<MyPositionMarkPoint> s_myPosition;
+static unique_ptr<PoiMarkPoint> s_selectionUserMark;
```
What is the point to use static keyword here, for global variables declared in anonymous namespace? Also, could you please rename them in to g_selectionUserMark and g_myPosition?

kw217 commented:
Much of this code is touched by clang-format not me! I can't fix all of the project's style! I will remove the clang-format since it is confusing people.

deathbaba commented:
The hint with clang-formatting only touched code should solve it.

kw217 commented:
Whew! I've applied all the markups above, fixing up a bunch of archaic C++ in the process. Hopefully this PR is in a better shape to merge now. Thanks for all the input - I've certainly learned some things.

JTTP

deathbaba commented:
LGTM
@gardster @ygorshenin @igrechuhin @ExMix @mpimenov PTAL please.

igrechuhin commented:
LGTM

ygorshenin commented on map/api_mark_container.cpp:
```diff
@@ -0,0 +1,18 @@
+#include "map/api_mark_container.hpp"
```
nit: it's better to group includes like there:
```c++
#include "map/api_mark_container.hpp"

#include "map/api_mark_point.hpp"
#include "map/framework.hpp"
```
because there is explicitly specified that this source contains implementation for api_mark_container.hpp.

ygorshenin commented on map/styled_point.cpp:
```diff
@@ -0,0 +1,56 @@
+}
+
+string GetDefaultStyle() { return kSupportedColors[0]; }
+
```
nit: useless blank line.

ygorshenin commented on map/styled_point.hpp:
```diff
@@ -0,0 +1,47 @@
+string GetDefaultStyle();
+}  // namespace style
+
+class StyledPoint : public ICustomDrawable
```
Is it possible to move StyledPoint to the style namespace?

kw217 replied:
I see why you're suggesting this, but won't it make it complicated to use? @deathbaba, what do you think?

deathbaba replied:
It's ok for me.

kw217 replied:
OK, I'll do that this evening.

ygorshenin commented on map/styled_point.cpp:
```diff
@@ -0,0 +1,56 @@
+  return GetStyle().empty() ? s_centre : s_offset;
+}
+
+namespace
```
Please, move that anonymous namespace to the top (after #include but before StyledPoint methods definition).

ygorshenin commented on map/bookmark.hpp:
```diff
@@ -96,7 +97,13 @@ class Bookmark : public ICustomDrawable
+  void SetType(string const & type)
+  {
+    m_data.SetType(type);
+    SetStyle(type);
```
StyledPoint::m_style duplicates BookmarkData::m_type. Why do we need this duplication? Is it possible to get rid of StyledPoint::m_style and introduce ApiMarkPoint::m_style instead?

ygorshenin replied:
For example, you can declare pure protected virtual method StyledPoint::GetStyle() and implement it in Bookmark and ApiMarkPoint. Boorkmark implementation will return style from m_data and ApiMarkPoint will return ApiMarkPoint::m_style.

kw217 replied:
Thanks, that's a definite improvement!

ygorshenin commented:
Sorry for the delay in review, just a few more comments.

kw217 commented:
Thanks, all markups applied except for the namespace change for StyledPoint. I'm happy to make that change too if others confirm it's a good idea.

kw217 commented:
Namespace now changed too. Hopefully ready to go?

deathbaba commented:
@ygorshenin PTAL & Merge please.

ygorshenin commented:
LGTM

mpimenov commented:
LGTM, I guess...
